CDCgov/prime-reportstream
# Session Analytics

This PR tracks a user's session length (in seconds) and upon idle or unload (closing browser in some form), sends that session length to Azure. 

A user session, as agreed upon with Chris, is the following:

- When a user opens ReportStream
- Ignores logged in / logged out state
- If a user swaps to another tab OR minimizes, pause the session
- Only when a user closes the browser OR idles long enough that our `useIdleTimer` kicks in, is a tracking event sent to Azure

Test Steps:
1. Since this is a draft PR, I left the envars in `frontend-react/.env.development` but you NEED to have `REACT_APP_APPLICATIONINSIGHTS_CONNECTION_STRING` set correctly to see events
2. Tool around on the site, and then if you add a `debugger` statement in `AppRouter` in the `onUnload` function, you can see the event as you close the page
3. You can also see the aggregate time, and how the time pausing works, if you just add a `console.log` of `sessionTimeAggregate` and see how it increases / pauses 

## Changes
The biggest set of changes is in `AppRouter` as we are adding two `eventListeners`, one on `visibilitychange` and one on `beforeunload`. I had to add this code to `AppRouter` as all the Context variables we're utilizing live in the `SessionContext` and the `AppRouter` is the first larger component that is instantiated AFTER `SessionContext` is.

## Checklist

### Testing

## Linked Issues
https://app.zenhub.com/workspaces/experience-607d9d5e68b95200150fec37/issues/gh/cdcgov/prime-reportstream/7627

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue
